torch/distributed/_functional_collectives.py

```diff
@@ -305,9 +321,7 @@ def reduce_scatter_tensor(
         self.size(0) % group_size == 0
     ), f"input dimension 0 ({self.size(0)} must be a multiple of group_size {group_size}"
     tensor = torch._C._nn.reduce_scatter_tensor(self, reduceOp, scatter_dim, tag, rankset, group_size)  # type: ignore[attr-defined]
-    res = AsyncCollectiveTensor(tensor)
-    _register_wrapper_tensor(res, tensor)
```

Comment:
i keep forgetting what the policy should be on calling _register...

we need to put a better comment here to immortalize it...

my current thought is
 - during eager we must register here
 - during tracing we never register here, but we require backends will register when calling the actual collective

Reply:
Backend's can optionally skip the _register_wrapper_tensor call if they emit a wait_tensor.

Reply:
hmmm can they? i thought wait_tensor will call into this code https://github.com/pytorch/pytorch/blob/master/torch/distributed/_functional_collectives.py#L96

which will find nothing has been registered, UNLESS the backend did the registration first

Reply:
well this is complicated.

in inductor, I made collectives call register and i made wait_tensor call wait_tensor.
* this way, if wait happens in a separate inductor graph or in eager, it still works

but if your backend knows that allreduce+wait will be in the same graph, it can emit more efficient code that lowers a collective to `work = dist.<collective>` and lowers wait_tensor to `work.wait()` which skips the need for ever registering.

Reply:
i guess the risk/concern is that if you aren't careful in backend design, wait_tensor op can become a silent no-op

Reply:
yeah, we can look into this sort of optimization later.